{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "import os\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.by import By\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "#Y40B10A.2a.1\n",
    "#F54H12.1c.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n    import pandas as pd\\n    import time\\n    import json\\n    import os\\n    from selenium import webdriver\\n    from selenium.webdriver.common.by import By\\n    import requests\\n    from bs4 import BeautifulSoup\\n\\n    #Y40B10A.2a.1\\n    #F54H12.1c.3\\n\\n    #json 靜態爬\\n    transcript_name =request.POST[\\'transcript_id\\'] #由transcript.js回傳\\n    response = requests.get(\\n            \"https://wormbase.org/rest/widget/transcript/\" + transcript_name + \"/sequences\",\\n            headers={\\n                \"User-Agent\": \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36\"\\n            })\\n\\n    soup = BeautifulSoup(response.text, \"html.parser\")\\n    data= json.loads(str(soup))\\n\\n    start = []\\n    stop = []\\n    #處理變成dict\\n    strand = data[\"fields\"][\\'unspliced_sequence_context\\'][\\'data\\'][\\'strand\\']\\n    #check + -\\n    if strand == \"+\" :\\n        data1 = data[\"fields\"][\\'unspliced_sequence_context\\'][\\'data\\'][\\'positive_strand\\'][\\'sequence\\']\\n        data2 = data[\"fields\"][\\'spliced_sequence_context\\'][\\'data\\'][\\'positive_strand\\'][\\'sequence\\']\\n        for i in range(0,len(data2)-1):\\n            if (data2[i].islower() & data2[i+1].isupper()):\\n                start.append(i+2)\\n            elif (data2[i].isupper() & data2[i+1].islower()):\\n                stop.append(i+1)\\n        if stop == []:\\n            stop.append(len(data2))\\n    else:\\n        data1 = data[\"fields\"][\\'unspliced_sequence_context\\'][\\'data\\'][\\'negative_strand\\'][\\'sequence\\']\\n        data2 = data[\"fields\"][\\'spliced_sequence_context\\'][\\'data\\'][\\'negative_strand\\'][\\'sequence\\']\\n        for i in range(0,len(data2)-1):\\n            if (data2[i].islower() & data2[i+1].isupper()):\\n                start.append(i+2)\\n            elif (data2[i].isupper() & data2[i+1].islower()):\\n                stop.append(i+1)\\n        if stop == []:\\n            stop.append(len(data2))\\n\\n    #靜態爬\\n\\n    #check + -\\n    if strand == \"+\" :\\n        transcriptdata1 = data[\"fields\"][\\'unspliced_sequence_context\\'][\\'data\\'][\\'positive_strand\\'][\\'features\\']\\n        transcriptdata2 = data[\"fields\"][\\'spliced_sequence_context\\'][\\'data\\'][\\'positive_strand\\'][\\'features\\']\\n       \\n    else:\\n        transcriptdata1= data[\"fields\"][\\'unspliced_sequence_context\\'][\\'data\\'][\\'negative_strand\\'][\\'features\\']\\n        transcriptdata2= data[\"fields\"][\\'spliced_sequence_context\\'][\\'data\\'][\\'negative_strand\\'][\\'features\\']\\n\\n    def count_len(start: int,end: int):\\n        return end-start +1\\n\\n    unspliced = pd.DataFrame(columns = [\\'type\\',\\'start\\',\\'stop\\'])\\n    spliced = pd.DataFrame(columns = [\\'type\\',\\'start\\',\\'stop\\'])\\n\\n    for i in range(len(transcriptdata1)):\\n        unspliced.loc[i] = transcriptdata1[i]\\n    \\n    for i in range(len(transcriptdata2)):\\n        spliced.loc[i] = transcriptdata2[i]\\n\\n\\n    CDS = {\\n    \\'type\\': \\'CDS\\',\\n    \\'start\\':start[len(start)-1],\\n    \\'stop\\':stop[len(stop)-1]}\\n\\n    CDS = pd.DataFrame([CDS])\\n    spliced = pd.concat([spliced, CDS], ignore_index=True)\\n\\n    unspliced[\\'length\\'] = unspliced.apply(lambda x: count_len(x[\\'start\\'],x[\\'stop\\']),axis=1)\\n    spliced[\\'length\\'] = spliced.apply(lambda x: count_len(x[\\'start\\'],x[\\'stop\\']),axis=1)\\n    \\n    #重新命名\\n    #unspliced \\n    count_Intron_u= 1\\n    count_Exon_u = 1\\n    for k in range(0,len(unspliced)):\\n        if unspliced.at[k,\\'type\\'] == \\'intron\\':\\n            unspliced.at[k,\\'type\\'] = \\'Intron\\' + str(count_Intron_u)\\n            count_Intron_u+= 1\\n        elif unspliced.at[k,\\'type\\'] == \\'exon\\':\\n            unspliced.at[k,\\'type\\'] = \\'Exon\\' + str(count_Exon_u)\\n            count_Exon_u += 1\\n        elif unspliced.at[k,\\'type\\'] == \\'five_prime_UTR\\':\\n            unspliced.at[k,\\'type\\'] = \\'5\\'UTR\\'\\n        elif unspliced.at[k,\\'type\\'] == \\'three_prime_UTR\\':\\n            unspliced.at[k,\\'type\\'] = \\'3\\'UTR\\'\\n    #spliced\\n    count_Exon_s = 1\\n    for k in range(0,len(spliced)):\\n        if spliced.at[k,\\'type\\'] == \\'exon\\':\\n            spliced.at[k,\\'type\\'] = \\'Exon\\' + str(count_Exon_s)\\n            count_Exon_s += 1\\n        elif spliced.at[k,\\'type\\'] == \\'five_prime_UTR\\':\\n            spliced.at[k,\\'type\\'] = \\'5\\'UTR\\'\\n        elif spliced.at[k,\\'type\\'] == \\'three_prime_UTR\\':\\n            spliced.at[k,\\'type\\'] = \\'3\\'UTR\\'\\n\\n    #for answer\\n    #unspliced       \\n    #print(\\'unspliced:\\n\\',unspliced)\\n    #spliced\\n    #print(\\'spliced:\\n\\',spliced)\\n    \\n\\n    translation = data[\\'fields\\'][\\'protein_sequence\\'][\\'data\\'][\\'sequence\\']\\n    #print(\\'translation:\\',translation)\\n             \\n'"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "    import pandas as pd\n",
    "    import time\n",
    "    import json\n",
    "    import os\n",
    "    from selenium import webdriver\n",
    "    from selenium.webdriver.common.by import By\n",
    "    import requests\n",
    "    from bs4 import BeautifulSoup\n",
    "\n",
    "    #Y40B10A.2a.1\n",
    "    #F54H12.1c.3\n",
    "\n",
    "    #json 靜態爬\n",
    "    transcript_name =request.POST['transcript_id'] #由transcript.js回傳\n",
    "    response = requests.get(\n",
    "            \"https://wormbase.org/rest/widget/transcript/\" + transcript_name + \"/sequences\",\n",
    "            headers={\n",
    "                \"User-Agent\": \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36\"\n",
    "            })\n",
    "\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    data= json.loads(str(soup))\n",
    "\n",
    "    start = []\n",
    "    stop = []\n",
    "    #處理變成dict\n",
    "    strand = data[\"fields\"]['unspliced_sequence_context']['data']['strand']\n",
    "    #check + -\n",
    "    if strand == \"+\" :\n",
    "        data1 = data[\"fields\"]['unspliced_sequence_context']['data']['positive_strand']['sequence']\n",
    "        data2 = data[\"fields\"]['spliced_sequence_context']['data']['positive_strand']['sequence']\n",
    "        for i in range(0,len(data2)-1):\n",
    "            if (data2[i].islower() & data2[i+1].isupper()):\n",
    "                start.append(i+2)\n",
    "            elif (data2[i].isupper() & data2[i+1].islower()):\n",
    "                stop.append(i+1)\n",
    "        if stop == []:\n",
    "            stop.append(len(data2))\n",
    "    else:\n",
    "        data1 = data[\"fields\"]['unspliced_sequence_context']['data']['negative_strand']['sequence']\n",
    "        data2 = data[\"fields\"]['spliced_sequence_context']['data']['negative_strand']['sequence']\n",
    "        for i in range(0,len(data2)-1):\n",
    "            if (data2[i].islower() & data2[i+1].isupper()):\n",
    "                start.append(i+2)\n",
    "            elif (data2[i].isupper() & data2[i+1].islower()):\n",
    "                stop.append(i+1)\n",
    "        if stop == []:\n",
    "            stop.append(len(data2))\n",
    "\n",
    "    #靜態爬\n",
    "\n",
    "    #check + -\n",
    "    if strand == \"+\" :\n",
    "        transcriptdata1 = data[\"fields\"]['unspliced_sequence_context']['data']['positive_strand']['features']\n",
    "        transcriptdata2 = data[\"fields\"]['spliced_sequence_context']['data']['positive_strand']['features']\n",
    "       \n",
    "    else:\n",
    "        transcriptdata1= data[\"fields\"]['unspliced_sequence_context']['data']['negative_strand']['features']\n",
    "        transcriptdata2= data[\"fields\"]['spliced_sequence_context']['data']['negative_strand']['features']\n",
    "\n",
    "    def count_len(start: int,end: int):\n",
    "        return end-start +1\n",
    "\n",
    "    unspliced = pd.DataFrame(columns = ['type','start','stop'])\n",
    "    spliced = pd.DataFrame(columns = ['type','start','stop'])\n",
    "\n",
    "    for i in range(len(transcriptdata1)):\n",
    "        unspliced.loc[i] = transcriptdata1[i]\n",
    "    \n",
    "    for i in range(len(transcriptdata2)):\n",
    "        spliced.loc[i] = transcriptdata2[i]\n",
    "\n",
    "\n",
    "    CDS = {\n",
    "    'type': 'CDS',\n",
    "    'start':start[len(start)-1],\n",
    "    'stop':stop[len(stop)-1]}\n",
    "\n",
    "    CDS = pd.DataFrame([CDS])\n",
    "    spliced = pd.concat([spliced, CDS], ignore_index=True)\n",
    "\n",
    "    unspliced['length'] = unspliced.apply(lambda x: count_len(x['start'],x['stop']),axis=1)\n",
    "    spliced['length'] = spliced.apply(lambda x: count_len(x['start'],x['stop']),axis=1)\n",
    "    \n",
    "    #重新命名\n",
    "    #unspliced \n",
    "    count_Intron_u= 1\n",
    "    count_Exon_u = 1\n",
    "    for k in range(0,len(unspliced)):\n",
    "        if unspliced.at[k,'type'] == 'intron':\n",
    "            unspliced.at[k,'type'] = 'Intron' + str(count_Intron_u)\n",
    "            count_Intron_u+= 1\n",
    "        elif unspliced.at[k,'type'] == 'exon':\n",
    "            unspliced.at[k,'type'] = 'Exon' + str(count_Exon_u)\n",
    "            count_Exon_u += 1\n",
    "        elif unspliced.at[k,'type'] == 'five_prime_UTR':\n",
    "            unspliced.at[k,'type'] = '5\\'UTR'\n",
    "        elif unspliced.at[k,'type'] == 'three_prime_UTR':\n",
    "            unspliced.at[k,'type'] = '3\\'UTR'\n",
    "    #spliced\n",
    "    count_Exon_s = 1\n",
    "    for k in range(0,len(spliced)):\n",
    "        if spliced.at[k,'type'] == 'exon':\n",
    "            spliced.at[k,'type'] = 'Exon' + str(count_Exon_s)\n",
    "            count_Exon_s += 1\n",
    "        elif spliced.at[k,'type'] == 'five_prime_UTR':\n",
    "            spliced.at[k,'type'] = '5\\'UTR'\n",
    "        elif spliced.at[k,'type'] == 'three_prime_UTR':\n",
    "            spliced.at[k,'type'] = '3\\'UTR'\n",
    "\n",
    "    #for answer\n",
    "    #unspliced       \n",
    "    #print('unspliced:\\n',unspliced)\n",
    "    #spliced\n",
    "    #print('spliced:\\n',spliced)\n",
    "    \n",
    "\n",
    "    translation = data['fields']['protein_sequence']['data']['sequence']\n",
    "    #print('translation:',translation)\n",
    "             \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n#for 動態爬蟲\\ntranscript_name = input(\"Enter the transcript name :\")\\nurl = \"https://wormbase.org/species/c_elegans/transcript/\" + transcript_name +\"#06--10\"\\ntime.sleep(2)\\n\\n#Y40B10A.2a.1\\n#F54H12.1c.3\\n\\n\\n# 動態crawler \\ndriver = webdriver.Chrome() \\ndriver.get(url)\\ntime.sleep(2)\\nbutton1 = driver.find_element(By.XPATH,\\'/html/body/div[2]/div[3]/div[3]/div[2]/div[1]/li[12]/div/div[2]/div[2]/div[1]/div[2]/div/div/div[1]/div/button[2]\\').click()\\nbutton2 = driver.find_element(By.XPATH,\\'/html/body/div[2]/div[3]/div[3]/div[2]/div[1]/li[12]/div/div[2]/div[2]/div[1]/div[2]/div/div/div[2]/div/button[2]\\').click()\\ntime.sleep(2)\\ndriver.close()\\n\\n\\n#unspliced\\nwith open(\\'/home/cosbi/Downloads/unspliced+UTRTranscriptSequence_\\' + transcript_name + \\'.fasta\\',\\'r\\') as file:\\n    data1 = file.readlines()\\n#spliced\\nwith open(\\'/home/cosbi/Downloads/spliced+UTRTranscriptSequence_\\' + transcript_name + \\'.fasta\\',\\'r\\') as file1:\\n    data2 = file1.readlines()\\n\\nprint(data1)\\nprint(data2)\\n'"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "#for 動態爬蟲\n",
    "transcript_name = input(\"Enter the transcript name :\")\n",
    "url = \"https://wormbase.org/species/c_elegans/transcript/\" + transcript_name +\"#06--10\"\n",
    "time.sleep(2)\n",
    "\n",
    "#Y40B10A.2a.1\n",
    "#F54H12.1c.3\n",
    "\n",
    "\n",
    "# 動態crawler \n",
    "driver = webdriver.Chrome() \n",
    "driver.get(url)\n",
    "time.sleep(2)\n",
    "button1 = driver.find_element(By.XPATH,'/html/body/div[2]/div[3]/div[3]/div[2]/div[1]/li[12]/div/div[2]/div[2]/div[1]/div[2]/div/div/div[1]/div/button[2]').click()\n",
    "button2 = driver.find_element(By.XPATH,'/html/body/div[2]/div[3]/div[3]/div[2]/div[1]/li[12]/div/div[2]/div[2]/div[1]/div[2]/div/div/div[2]/div/button[2]').click()\n",
    "time.sleep(2)\n",
    "driver.close()\n",
    "\n",
    "\n",
    "#unspliced\n",
    "with open('/home/cosbi/Downloads/unspliced+UTRTranscriptSequence_' + transcript_name + '.fasta','r') as file:\n",
    "    data1 = file.readlines()\n",
    "#spliced\n",
    "with open('/home/cosbi/Downloads/spliced+UTRTranscriptSequence_' + transcript_name + '.fasta','r') as file1:\n",
    "    data2 = file1.readlines()\n",
    "\n",
    "print(data1)\n",
    "print(data2)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aagagacgcagaaatcacaatgtttgtcatcgtttttttcttgttctctcggactatctgatagtataaattcttccctcttttgcctccccaatccaattaattttcagtttttttttccgaaaaaaaaaATGTCCAACGTCGCAAAAAGCTACAAGAGCTCGGATCCAGTGATTGCGTATTGCTCGGAGgtttgtagtttgtagtttgtagtttgtaatgtcacggatttctggtttccttcataaattgaaatggaagagtttgccgaactaggccattttggctcggccatatctgttgtagatttacggcgcgttgcgtgtcgcgtcgcggttcgagtttagttgttaaactaaatgtatttgtccgtgtggagtacacgactttcccacgcgttgtccggcgggcgattgtcaatggagcgcgaaaaattcaatgaggaaggcccacggtagcacagaaactagatgggttggtttatgatactcagtctggaaacctatattggctactatctccaaaaccatcacaaaatcgattttgcgatgcatataagaaatgaccgcaatgaaattatctatctttatttgtgatgaattttcgattttgtacttcctggccaagttatgcacgtttgttcgatggagcgcgtttgaacccatctagcaactgaaccaccgtgggccagaaccccccgtgaatcgtcgtggaaacgcgctccgcctccaacagttgggtctcactaggtaaaaatttgggcggcaaaattgtaaattcaaatttttttaccaattttcaagcattttcttgaagtttcagaagaaaaataaaataaacatttttttttcgatttttcaatattgcaaattcacataattctaaaactcgtacttaaaggcgcatggagttattcggatgggtctcaccgcgaaaaatgcccttttttacctcctccggtgctcgttttttatagggacgcagagcaaactcgccctaatgccaattttcgaagcttcgcccccgaaccatgggtctcgttaggtatttgggcggcaaaattgtgattacaaaggtttttcaccattccagttgttttttttttaattttttttgccgaaataacgcttaaaaatttatgtgaatttgaaaagtggaaaaagttttcacaaaaatcaatttcaatttcgcgttgcgtgtcgcgtcgcggctcaattttagttgcgaaactaaatgtatttgtccgtgtggagtacacgactttcccacgcgttgtccggcaggcgattgtcaatggagcgcgaaaaattcaaggaggaaggccagaaacccgtgtagtgcgtagcttatgaaaatattttttttcagcacacaacaatccaatcgccactacaaaatgattttttaaaattttattccttttttcgattttcacaacaaaaaaataaaaaattaatatttttaccacattattcgcccagttgcgaaatttagcgttttttggcccgaacatacgattaccggtctccaaacgaaattatgtatttaatgcaaaaaggcgtgtgacttcaaagggtactgtaacttttattgatttttctacacaatgtttttcatgatagactactaattcataaaactttgattaaacaggaaaaaatcaacgaaaaaacaatgaaattttcatattttccacgaaaaattagcgaaaaattacagcatttttgcatagaaaaatttattattaggaccccacatatttctggcttccctcataaattgaaatggaagagtttgccgaactaggccattttggcttggccatctggtgtagatttacggcgcgttgcgtgtcgcgtcgcggctcaattttagttgcgaaactaaacgtatttgtccgtgtggagtacacgactttcccatgtccggcaggcgattgtcaatggagcgcgaaaaattcaatgaggaaggccagaaacccgtgtagtgcgtagcttatgaaaatatttttattttctttttttccagCACACCACAATCCAATCGCCACTCCAGGCTGAGCTCCTCAAGGAGACACTCGACAATGCTCCAATGTCTATGATGCTCGGAGCGCCGGAAGTTCTCACGTTCGGACAGAATTTCATCAGATCGTTCGGTGGCAGGCGTGTCATTGATGTCGGTACCTACACCGGAGCATCTGCCTTGGCATGGGCTCTCGCTGTTCCGGATGATGGAGAAGTGTTCACTTTTGACATTGATCACACTAATTATAAAAAGTTCGGAGTTCCGATCATCTCGAAATGCGAGAAGACTTTCAAAAAGATTCGTGCTATCGAAGGATCCGCGTTGGATGGACTTGgtaagttgctcaattttttcggcagaaaaacttaaaaaattcaatacatacttcaagaatttcttgtgaaaaaacaatttttcgctttttttaaaagaaaaaaactagaaaacatgtaaaaaattattcaaaacacgtggtgacatagtgctccataacggtttgatctacgttgatctacgatattcgcgatttcgcatgaacgtgctgacgtcactttttgttgtgaaaaaattcccgcattttttgtagatcaaaccgtaatgggacagcctgacaccctttccaattactacgagaacacaatgatttgataattctgcgtactgcacaacatattttacgcgcaaaatatctcgttgcgaaaactacagtaatccttcaaatgacaactgtagcgcttgtgtgtcgatttacgggctcgattttcatttcaaaatcgagcccctgtaaatcgacacaagggctacagtagtcattaagaattactgtagttttcgctgcgagatattttgcgcgtcaaatgtgttgtaccgtaggcattctcggaattttgtgttcccgtaatatggcaacaactttttttttgcattttttcagaaaaaaaccaaaaaatcgatattttcagcgtttctggcaaaaaaaattttttttgctattttttcagaaatccgcagaaaaaccccggttttcaagaatttttcgattttttaaaattcaatttctcctaaaaacattttttttaaactaaaaatggctgattttgccgccaaattcccaacgagacccatgacttagtgccgtggagcgcgttttcacttgtaatattctagACAAGCTCATCGCCGACGGACAATCTGGAACCTTTGATTTTGCCTTCATCGATGCCGACAAGGTGAACTACAGTAACTATTACGATAGAGCTGTCACCCTGCTCCGCAAGGGAGGTGTCATTTTTGTAGACAATgtaagtagtggagcctgaattaaaaaaaaatttttttttctaaaatttcttgaattattcagAGTCTTTGGAACGGATCCGTGTGCGATCCAGCGAGAAGAACCGAGCCGAGCACCGTCGCGATCCACCAGGCAAATGACAAAATCTTCAAGGATGACCGTACCTATTCTGCTCTAATCAACCTTGGCGATGGAACTCATATCGCTTTTAAGAAGTGAattaggggctttttttttaattttgaattatatttaataaattttatcttttaaaa\n",
      "aagagacgcagaaatcacaatgtttgtcatcgtttttttcttgttctctcggactatctgatagtataaattcttccctcttttgcctccccaatccaattaattttcagtttttttttccgaaaaaaaaaATGTCCAACGTCGCAAAAAGCTACAAGAGCTCGGATCCAGTGATTGCGTATTGCTCGGAGCACACCACAATCCAATCGCCACTCCAGGCTGAGCTCCTCAAGGAGACACTCGACAATGCTCCAATGTCTATGATGCTCGGAGCGCCGGAAGTTCTCACGTTCGGACAGAATTTCATCAGATCGTTCGGTGGCAGGCGTGTCATTGATGTCGGTACCTACACCGGAGCATCTGCCTTGGCATGGGCTCTCGCTGTTCCGGATGATGGAGAAGTGTTCACTTTTGACATTGATCACACTAATTATAAAAAGTTCGGAGTTCCGATCATCTCGAAATGCGAGAAGACTTTCAAAAAGATTCGTGCTATCGAAGGATCCGCGTTGGATGGACTTGACAAGCTCATCGCCGACGGACAATCTGGAACCTTTGATTTTGCCTTCATCGATGCCGACAAGGTGAACTACAGTAACTATTACGATAGAGCTGTCACCCTGCTCCGCAAGGGAGGTGTCATTTTTGTAGACAATAGTCTTTGGAACGGATCCGTGTGCGATCCAGCGAGAAGAACCGAGCCGAGCACCGTCGCGATCCACCAGGCAAATGACAAAATCTTCAAGGATGACCGTACCTATTCTGCTCTAATCAACCTTGGCGATGGAACTCATATCGCTTTTAAGAAGTGAattaggggctttttttttaattttgaattatatttaataaattttatcttttaaaa\n"
     ]
    }
   ],
   "source": [
    "#json 靜態爬\n",
    "transcript_name = input(\"Enter the transcript name :\")\n",
    "response = requests.get(\n",
    "        \"https://wormbase.org/rest/widget/transcript/\" + transcript_name + \"/sequences\",\n",
    "        headers={\n",
    "            \"User-Agent\": \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36\"\n",
    "        })\n",
    "\n",
    "soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "data= json.loads(str(soup))\n",
    "\n",
    "#處理變成dict\n",
    "strand = data[\"fields\"]['unspliced_sequence_context']['data']['strand']\n",
    "#check + -\n",
    "if strand == \"+\" :\n",
    "    data1 = data[\"fields\"]['unspliced_sequence_context']['data']['positive_strand']['sequence']\n",
    "    data2 = data[\"fields\"]['spliced_sequence_context']['data']['positive_strand']['sequence']\n",
    "else:\n",
    "    data1 = data[\"fields\"]['unspliced_sequence_context']['data']['negative_strand']['sequence']\n",
    "    data2 = data[\"fields\"]['spliced_sequence_context']['data']['negative_strand']['sequence']\n",
    "print(data1)\n",
    "print(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = 0\n",
    "\n",
    "def check(a,b):\n",
    "    transcription = ''  # 空的list\n",
    "    i = 0\n",
    "    while i < len(a):\n",
    "        # 找Exon\n",
    "        if a[i].isupper():  # 判斷大寫\n",
    "            transcription += a[i]  # 把該個字元加進去\n",
    "            if i == len(a)-1:  # 用來停止while迴圈\n",
    "                break\n",
    "            elif a[i+1].islower():\n",
    "                transcription += ','\n",
    "        # 找Intron\n",
    "        elif a[i].islower():  # 判斷小寫\n",
    "            transcription += a[i]\n",
    "            if i == len(a)-1:\n",
    "                break\n",
    "            elif a[i+1].isupper():\n",
    "                transcription += ','\n",
    "        i += 1  # 逐個檢查\n",
    "    # 弄成DataFrame的形式\n",
    "    df1 = pd.DataFrame(columns=['Exon&Intron'])  # 建立兩個column的DataFrame\n",
    "    df1['Exon&Intron'] = transcription.split(',')\n",
    "    transcription = '' #空的list\n",
    "    i = 0 \n",
    "\n",
    "    while i <len(b):\n",
    "        #找Exon\n",
    "        if b[i].isupper(): #判斷大寫\n",
    "            transcription  += b[i] #把該個字元加進去\n",
    "            if i == len(b)-1: #用來停止while迴圈\n",
    "                break\n",
    "            elif b[i+1].islower():\n",
    "                transcription  += ','\n",
    "        #找Intron\n",
    "        elif b[i].islower(): #判斷小寫\n",
    "            transcription  += b[i]\n",
    "            if i == len(b)-1:\n",
    "                break\n",
    "            elif b[i+1].isupper():\n",
    "                transcription  += ','\n",
    "        i += 1 #逐個檢查\n",
    "    #弄成DataFrame的形式\n",
    "    df2 = pd.DataFrame(columns=['Exon&Intron']) #建立兩個column的DataFrame\n",
    "    df2['Exon&Intron'] = transcription.split(',')\n",
    "\n",
    "    global c\n",
    "    if df1.at[0,'Exon&Intron'] == df2.at[0,'Exon&Intron']:\n",
    "        c = 0\n",
    "        print('easy')\n",
    "\n",
    "    else:\n",
    "        c = 1 \n",
    "        print('hard')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy\n"
     ]
    }
   ],
   "source": [
    "#check 5'UTR 是否一樣\n",
    "check(data1,data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy\n",
      "unspliced:\n",
      "       Name   Start     End  Length\n",
      "0    5'UTR     1.0   131.0     131\n",
      "1    Exon1     1.0   191.0     191\n",
      "2  Intron1   192.0  2028.0    1837\n",
      "3    Exon2  2029.0  2359.0     331\n",
      "4  Intron2  2360.0  3192.0     833\n",
      "5    Exon3  3193.0  3326.0     134\n",
      "6  Intron3  3327.0  3388.0      62\n",
      "7    Exon4  3389.0  3600.0     212\n",
      "8    3'UTR  3545.0  3600.0      56\n",
      "spliced:\n",
      "     Name  Start    End  Length\n",
      "0  5'UTR    1.0  131.0     131\n",
      "1  Exon1    1.0  191.0     191\n",
      "2  Exon2  192.0  522.0     331\n",
      "3  Exon3  523.0  656.0     134\n",
      "4  Exon4  657.0  868.0     212\n",
      "5  3'UTR  813.0  868.0      56\n",
      "6    CDS  132.0  812.0     681\n",
      "translation: MSNVAKSYKSSDPVIAYCSEHTTIQSPLQAELLKETLDNAPMSMMLGAPEVLTFGQNFIRSFGGRRVIDVGTYTGASALAWALAVPDDGEVFTFDIDHTNYKKFGVPIISKCEKTFKKIRAIEGSALDGLDKLIADGQSGTFDFAFIDADKVNYSNYYDRAVTLLRKGGVIFVDNSLWNGSVCDPARRTEPSTVAIHQANDKIFKDDRTYSALINLGDGTHIAFKK\n"
     ]
    }
   ],
   "source": [
    "if c == 0:\n",
    "    print('easy')\n",
    "    #把大小寫切出來\n",
    "    \n",
    "    #for unspliced\n",
    "    transcription = '' #空的list\n",
    "    upper_lower = ''  #空的list 用來記錄有幾次大小寫變化\n",
    "    i = 0 \n",
    "\n",
    "    while i <len(data1):\n",
    "        #找Exon\n",
    "        if data1[i].isupper(): #判斷大寫\n",
    "            transcription  += data1[i] #把該個字元加進去\n",
    "            if i == len(data1)-1: #用來停止while迴圈\n",
    "                break\n",
    "            elif data1[i+1].islower():\n",
    "                transcription  += ','\n",
    "                upper_lower += 'U,'\n",
    "        #找Intron\n",
    "        elif data1[i].islower(): #判斷小寫\n",
    "            transcription  += data1[i]\n",
    "            if i == len(data1)-1:\n",
    "                break\n",
    "            elif data1[i+1].isupper():\n",
    "                transcription  += ','\n",
    "                upper_lower += 'L,'\n",
    "        i += 1 #逐個檢查\n",
    "    #處理最後一個的大小寫\n",
    "    if data1[len(data1)-2].isupper():\n",
    "        upper_lower +='U'\n",
    "    else:\n",
    "        upper_lower +='L'\n",
    "\n",
    "    #弄成DataFrame的形式\n",
    "    df1 = pd.DataFrame(columns=['temp','Exon&Intron']) #建立兩個column的DataFrame\n",
    "    df1['temp'] = upper_lower.split(',') #split 表示依照什麼東西來區別 因為是string\n",
    "    df1['Exon&Intron'] = transcription.split(',')\n",
    "\n",
    "    #處理5'UTR和3'UTR\n",
    "    if df1.at[0,'temp'] == 'L': #開頭小寫表示有5'UTR\n",
    "        df1.at[0,'temp'] = '5'\n",
    "    if df1.at[len(df1)-1,'temp'] == 'L': #結尾小寫表示有3'UTR\n",
    "        df1.at[len(df1)-1,'temp'] = '3'\n",
    "    #print(df1)\n",
    "    \n",
    "\n",
    "    #判斷Intron(小寫的)跟Exon(大寫的)\n",
    "    count_Intron= 1\n",
    "    count_Exon = 1\n",
    "    for k in range(0,len(df1)):\n",
    "        if df1.at[k,'temp'] == 'L':\n",
    "            df1.at[k,'Name'] = 'Intron' + str(count_Intron)\n",
    "            count_Intron+= 1\n",
    "        elif df1.at[k,'temp'] == 'U':\n",
    "            df1.at[k,'Name'] = 'Exon' + str(count_Exon)\n",
    "            count_Exon += 1\n",
    "        elif df1.at[k,'temp'] == '5':\n",
    "            df1.at[k,'Name'] = '5\\'UTR'\n",
    "        elif df1.at[k,'temp'] == '3':\n",
    "            df1.at[k,'Name'] = '3\\'UTR'\n",
    "    #計算Length\n",
    "    #弄一個dataframe 給spliced 用\n",
    "    df3 = pd.DataFrame()\n",
    "    df3 = df1\n",
    "    for k in range(0,len(df3)):\n",
    "        if df3.at[k,'temp'] == 'L' :\n",
    "            df3 = df3.drop(labels = k)\n",
    "    df3 = df3.reset_index()  \n",
    "    df3 = df3.drop(columns='index')\n",
    "\n",
    "    df1['Length'] = df1['Exon&Intron'].str.len()\n",
    "\n",
    "    #計算起始與終點位置方便標顏色\n",
    "    #處理頭\n",
    "    if df1.at[0,'temp'] != 'U' and df1.at[0,'temp'] !='L': #計算5'UTR\n",
    "        df1.at[0,'Start'] = 1 #新增column\n",
    "        df1.at[0,'End'] = df1.at[0,'Start'] + df1.at[0,'Length'] -1\n",
    "        #計算Exon1(因為要含5'utr所以比較特別)\n",
    "        df1.at[1,'Start'] = 1 \n",
    "        df1.at[1,'End'] = df1.at[0,'Length'] + df1.at[1,'Length'] \n",
    "        df1.at[1,'Length'] = df1.at[0,'Length'] + df1.at[1,'Length']\n",
    "    #處理沒有5'UTR的\n",
    "    if df1.at[0,'temp'] == 'U' :\n",
    "        df1.at[0,'Start'] = 1\n",
    "        df1.at[0,'End'] = df1.at[0,'Start'] +df1.at[0,'Length'] -1\n",
    "    #處理中間    \n",
    "    for l in range(0,len(df1)): #計算Exon & Intron\n",
    "        if df1.at[l,'Start'] !=1:\n",
    "            df1.at[l,'Start'] = df1.at[l-1,'End'] + 1 #第2個的開始由第1個的結尾\n",
    "            df1.at[l,'End'] = df1.at[l,'Start'] +df1.at[l,'Length'] -1\n",
    "    #處理尾         \n",
    "    if df1.at[len(df1)-1,'temp'] != 'U' and df1.at[len(df1)-1,'temp'] !='L': #計算3'UTR\n",
    "        df1.at[len(df1)-1,'End'] = df1.at[len(df1)-2,'End'] + df1.at[len(df1)-1,'Length']\n",
    "        df1.at[len(df1)-2,'End'] = df1.at[len(df1)-2,'End'] + df1.at[len(df1)-1,'Length'] #特別處理Exon因為要包含3'UTR\n",
    "        df1.at[len(df1)-2,'Length'] = df1.at[len(df1)-2,'Length'] + df1.at[len(df1)-1,'Length']\n",
    "\n",
    "    #for spliced\n",
    "    transcription_spliced = '' #空的list\n",
    "    upper_lower = ''  #空的list 用來記錄有幾次大小寫變化\n",
    "    i = 0 \n",
    "\n",
    "    while i <len(data2):\n",
    "        #找Exon\n",
    "        if data2[i].isupper(): #判斷大寫\n",
    "            transcription_spliced  += data2[i] #把該個字元加進去\n",
    "            if i == len(data2)-1: #用來停止while迴圈\n",
    "                break\n",
    "            elif data2[i+1].islower():\n",
    "                transcription_spliced  += ','\n",
    "                upper_lower += 'U,'\n",
    "        #找Intron\n",
    "        elif data2[i].islower(): #判斷小寫\n",
    "            transcription_spliced  += data2[i]\n",
    "            if i == len(data2)-1:\n",
    "                break\n",
    "            elif data2[i+1].isupper():\n",
    "                transcription_spliced  += ','\n",
    "                upper_lower += 'L,'\n",
    "        i += 1 #逐個檢查\n",
    "    #處理最後一個的大小寫\n",
    "    if data2[len(data2)-2].isupper():\n",
    "        upper_lower +='U'\n",
    "    else:\n",
    "        upper_lower +='L'\n",
    "\n",
    "\n",
    "    #弄成DataFrame的形式\n",
    "    df2 = pd.DataFrame(columns=['temp','Exon&Intron']) #建立兩個column的DataFrame\n",
    "    df2['temp'] = upper_lower.split(',') #split 表示依照什麼東西來區別 因為是string\n",
    "    df2['Exon&Intron'] = transcription_spliced.split(',')\n",
    "\n",
    "    #處理5'UTR和3'UTR\n",
    "    if df2.at[0,'temp'] == 'L': #開頭小寫表示有5'UTR\n",
    "        df2.at[0,'temp'] = '5'\n",
    "    \n",
    "    if df2.at[len(df2)-1,'temp'] == 'L': #結尾小寫表示有3'UTR\n",
    "        df2.at[len(df2)-1,'temp'] = '3'\n",
    "    \n",
    "    #刪掉Intron\n",
    "\n",
    "    df3['Length'] = df3['Exon&Intron'].str.len()\n",
    "\n",
    "    #計算起始與終點位置方便標顏色\n",
    "    #處理頭\n",
    "    if df3.at[0,'temp'] != 'U' and df3.at[0,'temp'] !='L': #計算5'UTR\n",
    "        df3.at[0,'Start'] = 1 #新增column\n",
    "        df3.at[0,'End'] = df3.at[0,'Start'] + df3.at[0,'Length'] -1\n",
    "        #計算Exon1(因為要含5'utr所以比較特別)\n",
    "        df3.at[1,'Start'] = 1 \n",
    "        df3.at[1,'End'] = df3.at[0,'Length'] + df3.at[1,'Length'] \n",
    "        df3.at[1,'Length'] = df3.at[0,'Length'] + df3.at[1,'Length']\n",
    "    #處理沒有5'UTR的\n",
    "    if df3.at[0,'temp'] == 'U' :\n",
    "        df3.at[0,'Start'] = 1\n",
    "        df3.at[0,'End'] = df3.at[0,'Start'] +df3.at[0,'Length'] -1\n",
    "    #處理中間    \n",
    "    for l in range(0,len(df3)): #計算Exon & Intron\n",
    "        if df3.at[l,'Start'] !=1:\n",
    "            df3.at[l,'Start'] = df3.at[l-1,'End'] + 1 #第2個的開始由第1個的結尾\n",
    "            df3.at[l,'End'] = df3.at[l,'Start'] +df3.at[l,'Length'] -1\n",
    "    #處理尾         \n",
    "    if df3.at[len(df3)-1,'temp'] != 'U' and df3.at[len(df3)-1,'temp'] !='L': #計算3'UTR\n",
    "        df3.at[len(df3)-1,'End'] = df3.at[len(df3)-2,'End'] + df3.at[len(df3)-1,'Length']\n",
    "        df3.at[len(df3)-2,'End'] = df3.at[len(df3)-2,'End'] + df3.at[len(df3)-1,'Length'] #特別處理Exon因為要包含3'UTR\n",
    "        df3.at[len(df3)-2,'Length'] = df3.at[len(df3)-2,'Length'] + df3.at[len(df3)-1,'Length']\n",
    "    #CDS\n",
    "    CDS = {\n",
    "        'temp': 'CDS',\n",
    "        'Exon&Intron': df2.at[1,'Exon&Intron'],\n",
    "        'Name': 'CDS',\n",
    "        'Length':len(df2.at[1,'Exon&Intron']),\n",
    "        'Start':len(df1.at[0,'Exon&Intron'])+1,\n",
    "        'End':len(df2.at[0,'Exon&Intron'])+len(df2.at[1,'Exon&Intron'])}\n",
    "    \n",
    "    df4 = pd.DataFrame()\n",
    "    df4 = df3\n",
    "    CDS = pd.DataFrame([CDS])\n",
    "    df3 = pd.concat([df3, CDS], ignore_index=True)\n",
    "\n",
    "\n",
    "    specific_row_index = 1  # Replace with the desired index\n",
    "    # Insert CDS at the specified row position in df3\n",
    "    df4 = pd.concat([df4.iloc[:specific_row_index], CDS, df4.iloc[specific_row_index:]], ignore_index=True)\n",
    "    \n",
    "    position_to_insert = 2\n",
    "    # part1 = df4.iloc[:position_to_insert]\n",
    "    # part2 = df4.iloc[position_to_insert+1:]\n",
    "    for i in range(0,len(df4)):\n",
    "        if (df4.at[i,'Name'] == '3\\'UTR'):\n",
    "            row_to_insert = df4.iloc[i]\n",
    "            df4 = pd.concat([df4.iloc[:position_to_insert], row_to_insert.to_frame().T, df4.iloc[position_to_insert:]], ignore_index=True)\n",
    "            df4 = df4.drop(i)\n",
    "    spliced_sp = ['Name','Start','End','Length']       \n",
    "    spliced_sp = df4[spliced_sp]\n",
    "    \n",
    "    #for answer\n",
    "    #unspliced\n",
    "    unspliced =['Name','Start','End','Length']\n",
    "    unspliced = df1[unspliced]\n",
    "    print('unspliced:\\n',unspliced)\n",
    "    #spliced\n",
    "    spliced = ['Name','Start','End','Length']\n",
    "    spliced = df3[spliced]\n",
    "    print('spliced:\\n',spliced)\n",
    "    \n",
    "    #處理Codon\n",
    "    a=df2.at[len(df2)-2,'Exon&Intron']\n",
    "    i=0\n",
    "    codon=''\n",
    "    for k in range(0,len(a)):\n",
    "        if a[k].isupper():\n",
    "            codon += a[k]\n",
    "            i += 1\n",
    "            if (i%3) == 0 :\n",
    "                codon += ','\n",
    "            else:\n",
    "                continue    \n",
    "    #string to list\n",
    "    codon = codon.split(',')\n",
    "    del codon[len(codon)-1]\n",
    "     \n",
    "    #the dict of codon unspliced\n",
    "    dict = {'TTT':'F', 'TTC':'F','TTA':'L','TTG':'L','CTT':'L','CTC':'L','CTA':'L','CTG':'L','ATT':'I','ATC':'I','ATA':'I','ATG':'M','GTT':'V','GTC':'V','GTA':'V','GTG':'V',\n",
    "    'TCT':'S','TCC':'S','TCA':'S','TCG':'S','CCT':'P','CCC':'P','CCA':'P','CCG':'P','ACT':'T','ACC':'T','ACA':'T','ACG':'T','GCT':'A','GCC':'A','GCA':'A','GCG':'A','TAT':'Y','TAC':'Y','CAT':'H','CAC':'H','CAA':'Q','CAG':'Q','AAT':'N','AAC':'N','AAA':'K','AAG':'K','GAT':'D',\n",
    "    'GAC':'D','GAA':'E','GAG':'E','TGT':'C','TGC':'C','TGG':'W','CGT':'R','CGC':'R','CGA':'R','CGG':'R','AGT':'S','AGC':'S','AGA':'R','AGG':'R','GGT':'G','GGC':'G','GGA':'G','GGG':'G','TAA':'','TGA':'','TAG':''}\n",
    "\n",
    "    #output\n",
    "    Stop = 0\n",
    "    translation= []\n",
    "    for values in codon :\n",
    "        gene = dict.get(values)\n",
    "\n",
    "        if Stop == 0:\n",
    "            translation += gene        \n",
    "        if gene == '' :\n",
    "            Stop = 1         \n",
    "        elif Stop == 1:\n",
    "            print('end')\n",
    "            Stop = 0\n",
    "            break\n",
    "    translation = ''.join(translation)\n",
    "    print('translation:',translation)\n",
    "else:\n",
    "    print('hard')\n",
    "    '''\n",
    "    #靜態爬\n",
    "    response = requests.get(\n",
    "        \"https://wormbase.org/rest/widget/transcript/F54H12.1c.3/sequences\",\n",
    "        headers={\n",
    "            \"User-Agent\": \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36\"\n",
    "        })\n",
    "\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    data= json.loads(str(soup))\n",
    "    #處理變成dict\n",
    "    strand = data[\"fields\"]['unspliced_sequence_context']['data']['strand']\n",
    "    '''\n",
    "    #check + -\n",
    "    if strand == \"+\" :\n",
    "        transcriptdata1 = data[\"fields\"]['unspliced_sequence_context']['data']['positive_strand']['features']\n",
    "        transcriptdata2 = data[\"fields\"]['spliced_sequence_context']['data']['positive_strand']['features']\n",
    "    else:\n",
    "        transcriptdata1= data[\"fields\"]['unspliced_sequence_context']['data']['negative_strand']['features']\n",
    "        transcriptdata2= data[\"fields\"]['spliced_sequence_context']['data']['negative_strand']['features']\n",
    "\n",
    "    def count_len(start: int,end: int):\n",
    "        return end-start +1\n",
    "\n",
    "    unspliced = pd.DataFrame(columns = ['type','start','stop'])\n",
    "    spliced = pd.DataFrame(columns = ['type','start','stop'])\n",
    "\n",
    "    for i in range(len(transcriptdata1)):\n",
    "        unspliced.loc[i] = transcriptdata1[i]\n",
    "    \n",
    "    for i in range(len(transcriptdata2)):\n",
    "        spliced.loc[i] = transcriptdata2[i]\n",
    "\n",
    "\n",
    "    unspliced['length'] = unspliced.apply(lambda x: count_len(x['start'],x['stop']),axis=1)\n",
    "    spliced['length'] = spliced.apply(lambda x: count_len(x['start'],x['stop']),axis=1)\n",
    "    \n",
    "    #重新命名\n",
    "    #unspliced \n",
    "    count_Intron_u= 1\n",
    "    count_Exon_u = 1\n",
    "    for k in range(0,len(unspliced)):\n",
    "        if unspliced.at[k,'type'] == 'intron':\n",
    "            unspliced.at[k,'type'] = 'Intron' + str(count_Intron_u)\n",
    "            count_Intron_u+= 1\n",
    "        elif unspliced.at[k,'type'] == 'exon':\n",
    "            unspliced.at[k,'type'] = 'Exon' + str(count_Exon_u)\n",
    "            count_Exon_u += 1\n",
    "        elif unspliced.at[k,'type'] == 'five_prime_UTR':\n",
    "            unspliced.at[k,'type'] = '5\\'UTR'\n",
    "        elif unspliced.at[k,'type'] == 'three_prime_UTR':\n",
    "            unspliced.at[k,'type'] = '3\\'UTR'\n",
    "    #spliced\n",
    "    count_Exon_s = 1\n",
    "    for k in range(0,len(spliced)):\n",
    "        if spliced.at[k,'type'] == 'exon':\n",
    "            spliced.at[k,'type'] = 'Exon' + str(count_Exon_s)\n",
    "            count_Exon_s += 1\n",
    "        elif spliced.at[k,'type'] == 'five_prime_UTR':\n",
    "            spliced.at[k,'type'] = '5\\'UTR'\n",
    "        elif spliced.at[k,'type'] == 'three_prime_UTR':\n",
    "            spliced.at[k,'type'] = '3\\'UTR'\n",
    "  \n",
    "    #for answer\n",
    "    #unspliced       \n",
    "    print('unspliced:\\n',unspliced)\n",
    "\n",
    "    #spliced\n",
    "    print('spliced:\\n',spliced)\n",
    "    \n",
    "\n",
    "    translation = data['fields']['protein_sequence']['data']['sequence']\n",
    "    print('translation:',translation)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Name  Start    End Length\n",
      "0  5'UTR    1.0  131.0    131\n",
      "1    CDS  132.0  812.0    681\n",
      "2  3'UTR  813.0  868.0     56\n",
      "3  Exon1    1.0  191.0    191\n",
      "4  Exon2  192.0  522.0    331\n",
      "5  Exon3  523.0  656.0    134\n",
      "6  Exon4  657.0  868.0    212\n",
      "7  3'UTR  813.0  868.0     56\n"
     ]
    }
   ],
   "source": [
    "print(spliced_sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [],
   "source": [
    "unspliced = unspliced.values.tolist()\n",
    "spliced = spliced.values.tolist()\n",
    "\n",
    "unspliced_list = []\n",
    "spliced_list = []\n",
    "translation_list = []\n",
    "\n",
    "\n",
    "\n",
    "for i in range(len(unspliced)):\n",
    "        unspliced_list.append({\"Name\":unspliced[i][0]})\n",
    "        unspliced_list[i][\"Start\"] = unspliced[i][1]\n",
    "        unspliced_list[i][\"End\"] = unspliced[i][2]\n",
    "        unspliced_list[i][\"Length\"] = unspliced[i][3]\n",
    "        \n",
    "\n",
    "for i in range(len(spliced)):\n",
    "        spliced_list.append({\"Name\":spliced[i][0]})\n",
    "        spliced_list[i][\"Start\"] = spliced[i][1]\n",
    "        spliced_list[i][\"End\"] = spliced[i][2]\n",
    "        spliced_list[i][\"Length\"] = spliced[i][3]\n",
    "\n",
    "translation_list.append({\"translation\":translation}) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]\n",
      "3600\n",
      "543\n"
     ]
    }
   ],
   "source": [
    "unspliced_index = []\n",
    "exon_type_u = 0\n",
    "if ('3\\'UTR' in unspliced):\n",
    "    if ('5\\'UTR' in unspliced):\n",
    "        print('1')\n",
    "        for i in range(0,len(unspliced)):\n",
    "            if (unspliced[i][0] == '5\\'UTR'):\n",
    "                for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                    unspliced_index.append(1)\n",
    "            elif('Exon' in unspliced[i][0]):\n",
    "                    if(exon_type_u == 0 ) :\n",
    "                        if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1]-1)):\n",
    "                                unspliced_index.append(2)\n",
    "                                exon_type_u = 1 \n",
    "                        else:\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                                unspliced_index.append(2)\n",
    "                                exon_type_u = 1\n",
    "                    elif (exon_type_u == 1) :\n",
    "                        if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1]-1)):\n",
    "                                unspliced_index.append(3) \n",
    "                                exon_type_u =0\n",
    "                        else:\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                                unspliced_index.append(3)\n",
    "                                exon_type_u =0                \n",
    "            elif('Intron' in unspliced[i][0]):\n",
    "                    if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1])):\n",
    "                            unspliced_index.append(4) \n",
    "                    else:\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                            unspliced_index.append(4)\n",
    "            elif(unspliced[i][0] == '3\\'UTR'):\n",
    "                for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                    unspliced_index.append(5)\n",
    "    else:\n",
    "        print('2')\n",
    "        for i in range(0,len(unspliced)):\n",
    "            if('Exon' in unspliced[i][0]):\n",
    "                    if(exon_type_u == 0 ) :\n",
    "                        if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1]-1)):\n",
    "                                unspliced_index.append(2)\n",
    "                                exon_type_u = 1 \n",
    "                        else:\n",
    "                            for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                                unspliced_index.append(2)\n",
    "                                exon_type_u = 1\n",
    "                    elif (exon_type_u == 1) :\n",
    "                        if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1]-1)):\n",
    "                                unspliced_index.append(3) \n",
    "                                exon_type_u =0\n",
    "                        else:\n",
    "                            for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                                unspliced_index.append(3)\n",
    "                                exon_type_u =0                \n",
    "            elif('Intron' in unspliced[i][0]):\n",
    "                    if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1])):\n",
    "                            unspliced_index.append(4) \n",
    "                    else:\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                            unspliced_index.append(4)\n",
    "            elif(unspliced[i][0] == '3\\'UTR'):\n",
    "                for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                    unspliced_index.append(5)\n",
    "else:\n",
    "    \n",
    "    if ('5\\'UTR' in unspliced):\n",
    "        print('3')\n",
    "        for i in range(0,len(unspliced)):\n",
    "            if (unspliced[i][0] == '5\\'UTR'):\n",
    "                for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                    unspliced_index.append(1)\n",
    "            elif('Exon' in unspliced[i][0]):\n",
    "                    if(exon_type_u == 0 ) :\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                            unspliced_index.append(2)\n",
    "                            exon_type_u = 1\n",
    "                    elif (exon_type_u == 1) :\n",
    "                        for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                            unspliced_index.append(3)\n",
    "                            exon_type_u =0                \n",
    "            elif('Intron' in unspliced[i][0]):\n",
    "                    # if(unspliced[i+1][0] == '3\\'UTR'):\n",
    "                    #     for j in range(int(unspliced[i-1][2]),int(unspliced[i+1][1])):\n",
    "                    #         unspliced_index.append(4) \n",
    "                    # else:\n",
    "                    for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                        unspliced_index.append(4)\n",
    "            # elif(unspliced[i][0] == '3\\'UTR'):\n",
    "            #     for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "            #         unspliced_index.append(5)\n",
    "    else:\n",
    "        print('4')\n",
    "        for i in range(0,len(unspliced)):\n",
    "            if('Exon' in unspliced[i][0]):\n",
    "                if(exon_type_u == 0 ) :\n",
    "                    for j in range(int(unspliced[i][1]-1),int(unspliced[i][2])):\n",
    "                        unspliced_index.append(2)\n",
    "                        exon_type_u = 1\n",
    "                elif (exon_type_u == 1) :\n",
    "                    for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                        unspliced_index.append(3)\n",
    "                        exon_type_u =0                \n",
    "            elif('Intron' in unspliced[i][0]):\n",
    "                for j in range(int(unspliced[i-1][2]),int(unspliced[i][2])):\n",
    "                    unspliced_index.append(4)\n",
    "\n",
    "print(unspliced_index)\n",
    "print(len(unspliced_index))\n",
    "print(unspliced_index.count(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]\n",
      "868\n",
      "543\n"
     ]
    }
   ],
   "source": [
    "spliced_index = []\n",
    "exon_type_s = 0\n",
    "if ('3\\'UTR' in unspliced):\n",
    "    if ('5\\'UTR' in unspliced):\n",
    "        for i in range(0,len(spliced)):\n",
    "            if (spliced[i][0] == '5\\'UTR'):\n",
    "                for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                    spliced_index.append(1)\n",
    "            elif('Exon' in spliced[i][0]):\n",
    "                    if(exon_type_s == 0 ) :\n",
    "                        if(spliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(spliced[i-1][2]),int(spliced[i+1][1]-1)):\n",
    "                                spliced_index.append(2)\n",
    "                                exon_type_s = 1 \n",
    "                        else:\n",
    "                            for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                                spliced_index.append(2)\n",
    "                                exon_type_s = 1\n",
    "                    elif (exon_type_s == 1) :\n",
    "                        if(spliced[i+1][0] == '3\\'UTR'):\n",
    "                            for j in range(int(spliced[i-1][2]),int(spliced[i+1][1]-1)):\n",
    "                                spliced_index.append(3) \n",
    "                                exon_type_s =0\n",
    "                        else:\n",
    "                            for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                                spliced_index.append(3)\n",
    "                                exon_type_s =0                \n",
    "            elif(spliced[i][0] == '3\\'UTR'):\n",
    "                for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                    spliced_index.append(5)\n",
    "    else:\n",
    "        for i in range(0,len(spliced)):\n",
    "            if('Exon' in spliced[i][0]):\n",
    "                if(exon_type_s == 0 ) :\n",
    "                    if(spliced[i+1][0] == '3\\'UTR'):\n",
    "                        for j in range(int(spliced[i-1][2]),int(spliced[i+1][1]-1)):\n",
    "                            spliced_index.append(2)\n",
    "                            exon_type_s = 1 \n",
    "                    else:\n",
    "                        for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                            spliced_index.append(2)\n",
    "                            exon_type_s = 1\n",
    "                elif (exon_type_s == 1) :\n",
    "                    if(spliced[i+1][0] == '3\\'UTR'):\n",
    "                        for j in range(int(spliced[i-1][2]),int(spliced[i+1][1]-1)):\n",
    "                            spliced_index.append(3) \n",
    "                            exon_type_s =0\n",
    "                    else:\n",
    "                        for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                            spliced_index.append(3)\n",
    "                            exon_type_s =0                \n",
    "            elif(spliced[i][0] == '3\\'UTR'):\n",
    "                for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                    spliced_index.append(5)\n",
    "else:\n",
    "    if ('5\\'UTR' in unspliced):\n",
    "        for i in range(0,len(spliced)):\n",
    "            if (spliced[i][0] == '5\\'UTR'):\n",
    "                for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                    spliced_index.append(1)\n",
    "            elif('Exon' in spliced[i][0]):\n",
    "                    if(exon_type_s == 0 ) :\n",
    "                        for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                            spliced_index.append(2)\n",
    "                            exon_type_s = 1\n",
    "                    elif (exon_type_s == 1) :\n",
    "                        for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                            spliced_index.append(3)\n",
    "                            exon_type_s =0                \n",
    "    else:\n",
    "        for i in range(0,len(spliced)):\n",
    "            if('Exon' in spliced[i][0]):\n",
    "                if(exon_type_s == 0 ) :\n",
    "                    for j in range(int(spliced[i][1]-1),int(spliced[i][2])):\n",
    "                        spliced_index.append(2)\n",
    "                        exon_type_s = 1\n",
    "                elif (exon_type_s == 1) :\n",
    "                    for j in range(int(spliced[i-1][2]),int(spliced[i][2])):\n",
    "                        spliced_index.append(3)\n",
    "                        exon_type_s =0     \n",
    "print(spliced_index)\n",
    "print(len(spliced_index))\n",
    "print(spliced_index.count(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[\"5'UTR\", 1.0, 131.0, 131], ['Exon1', 1.0, 191.0, 191], ['Intron1', 192.0, 2028.0, 1837], ['Exon2', 2029.0, 2359.0, 331], ['Intron2', 2360.0, 3192.0, 833], ['Exon3', 3193.0, 3326.0, 134], ['Intron3', 3327.0, 3388.0, 62], ['Exon4', 3389.0, 3600.0, 212], [\"3'UTR\", 3545.0, 3600.0, 56]]\n",
      "[[\"5'UTR\", 1.0, 131.0, 131], ['Exon1', 1.0, 191.0, 191], ['Exon2', 192.0, 522.0, 331], ['Exon3', 523.0, 656.0, 134], ['Exon4', 657.0, 868.0, 212], [\"3'UTR\", 813.0, 868.0, 56], ['CDS', 132.0, 812.0, 681]]\n",
      "no\n"
     ]
    }
   ],
   "source": [
    "print(unspliced)\n",
    "print(spliced)\n",
    "if ('3\\'UTR' in unspliced):\n",
    "    print('yes')\n",
    "else:\n",
    "    print('no')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [],
   "source": [
    "unspliced_d3 = []\n",
    "exon_type_d3_u = 0\n",
    "for i in range(0,len(unspliced)):\n",
    "    if (unspliced[i][0] == '5\\'UTR'):\n",
    "        unspliced_d3.append({\"Type\":'5\\'UTR'})\n",
    "        unspliced_d3[i][\"start\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"end\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"x1\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"y1\"] = 50\n",
    "        unspliced_d3[i][\"x2\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"y2\"] = 50\n",
    "        unspliced_d3[i][\"color\"] = \"gray\"\n",
    "    elif('Exon' in unspliced[i][0]):\n",
    "            if(exon_type_d3_u == 0 ) :\n",
    "                unspliced_d3.append({\"Type\":'Exon'})\n",
    "                unspliced_d3[i][\"start\"] = unspliced[i][1]\n",
    "                unspliced_d3[i][\"end\"] = unspliced[i][2]\n",
    "                unspliced_d3[i][\"x1\"] = unspliced[i][1]\n",
    "                unspliced_d3[i][\"y1\"] = 50\n",
    "                unspliced_d3[i][\"x2\"] = unspliced[i][2]\n",
    "                unspliced_d3[i][\"y2\"] = 50\n",
    "                unspliced_d3[i][\"color\"] = \"yellow\"\n",
    "                exon_type_d3_u = 1\n",
    "            elif (exon_type_d3_u == 1) :\n",
    "                unspliced_d3.append({\"Type\":'Exon'})\n",
    "                unspliced_d3[i][\"start\"] = unspliced[i][1]\n",
    "                unspliced_d3[i][\"end\"] = unspliced[i][2]\n",
    "                unspliced_d3[i][\"x1\"] = unspliced[i][1]\n",
    "                unspliced_d3[i][\"y1\"] = 50\n",
    "                unspliced_d3[i][\"x2\"] = unspliced[i][2]\n",
    "                unspliced_d3[i][\"y2\"] = 50\n",
    "                unspliced_d3[i][\"color\"] = \"orange\" \n",
    "                exon_type_d3_u = 0             \n",
    "    elif('Intron' in unspliced[i][0]):\n",
    "        unspliced_d3.append({\"Type\":'Intron'})\n",
    "        unspliced_d3[i][\"start\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"end\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"x1\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"y1\"] = 50\n",
    "        unspliced_d3[i][\"x2\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"y2\"] = 50\n",
    "        unspliced_d3[i][\"color\"] = \"GhostWhite\"          \n",
    "    elif(unspliced[i][0] == '3\\'UTR'):\n",
    "        unspliced_d3.append({\"Type\":'3\\'UTR'})\n",
    "        unspliced_d3[i][\"start\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"end\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"x1\"] = unspliced[i][1]\n",
    "        unspliced_d3[i][\"y1\"] = 50\n",
    "        unspliced_d3[i][\"x2\"] = unspliced[i][2]\n",
    "        unspliced_d3[i][\"y2\"] = 50\n",
    "        unspliced_d3[i][\"color\"] = \"gray\"          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Type': \"5'UTR\", 'start': 1.0, 'end': 131.0, 'x1': 1.0, 'y1': 50, 'x2': 131.0, 'y2': 50, 'color': 'gray'}, {'Type': 'Exon', 'start': 1.0, 'end': 191.0, 'x1': 1.0, 'y1': 50, 'x2': 191.0, 'y2': 50, 'color': 'yellow'}, {'Type': 'Intron', 'start': 192.0, 'end': 2028.0, 'x1': 192.0, 'y1': 50, 'x2': 2028.0, 'y2': 50, 'color': 'GhostWhite'}, {'Type': 'Exon', 'start': 2029.0, 'end': 2359.0, 'x1': 2029.0, 'y1': 50, 'x2': 2359.0, 'y2': 50, 'color': 'orange'}, {'Type': 'Intron', 'start': 2360.0, 'end': 3192.0, 'x1': 2360.0, 'y1': 50, 'x2': 3192.0, 'y2': 50, 'color': 'GhostWhite'}, {'Type': 'Exon', 'start': 3193.0, 'end': 3326.0, 'x1': 3193.0, 'y1': 50, 'x2': 3326.0, 'y2': 50, 'color': 'yellow'}, {'Type': 'Intron', 'start': 3327.0, 'end': 3388.0, 'x1': 3327.0, 'y1': 50, 'x2': 3388.0, 'y2': 50, 'color': 'GhostWhite'}, {'Type': 'Exon', 'start': 3389.0, 'end': 3600.0, 'x1': 3389.0, 'y1': 50, 'x2': 3600.0, 'y2': 50, 'color': 'orange'}, {'Type': \"3'UTR\", 'start': 3545.0, 'end': 3600.0, 'x1': 3545.0, 'y1': 50, 'x2': 3600.0, 'y2': 50, 'color': 'gray'}]\n"
     ]
    }
   ],
   "source": [
    "print(unspliced_d3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [],
   "source": [
    "spliced_d3 = []\n",
    "exon_type_d3_s = 0\n",
    "for i in range(0,len(spliced)):\n",
    "    if (spliced[i][0] == '5\\'UTR'):\n",
    "        spliced_d3.append({\"x1\":(spliced[i][1]/10)+360})\n",
    "        spliced_d3[i][\"y1\"] = 50\n",
    "        spliced_d3[i][\"x2\"] = (spliced[i][2]/10)+360\n",
    "        spliced_d3[i][\"y2\"] = 50\n",
    "        spliced_d3[i][\"color\"] = \"gray\"\n",
    "    elif('Exon' in spliced[i][0]):\n",
    "            if(exon_type_d3_s == 0 ) :\n",
    "                spliced_d3.append({\"x1\":(spliced[i][1]/10)+360})\n",
    "                spliced_d3[i][\"y1\"] = 50\n",
    "                spliced_d3[i][\"x2\"] = (spliced[i][2]/10)+360\n",
    "                spliced_d3[i][\"y2\"] = 50\n",
    "                spliced_d3[i][\"color\"] = \"yellow\"\n",
    "                exon_type_d3_s = 1\n",
    "            elif (exon_type_d3_s == 1) :\n",
    "                spliced_d3.append({\"x1\":(spliced[i][1]/10)+360})\n",
    "                spliced_d3[i][\"y1\"] = 50\n",
    "                spliced_d3[i][\"x2\"] = (spliced[i][2]/10)+360\n",
    "                spliced_d3[i][\"y2\"] = 50\n",
    "                spliced_d3[i][\"color\"] = \"orange\" \n",
    "                exon_type_d3_s = 0               \n",
    "    elif(spliced[i][0] == '3\\'UTR'):\n",
    "        spliced_d3.append({\"x1\":(spliced[i][1]/10)+360})\n",
    "        spliced_d3[i][\"y1\"] = 50\n",
    "        spliced_d3[i][\"x2\"] = (spliced[i][2]/10)+360\n",
    "        spliced_d3[i][\"y2\"] = 50\n",
    "        spliced_d3[i][\"color\"] = \"gray\"  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'x1': 360.1, 'y1': 50, 'x2': 373.1, 'y2': 50, 'color': 'gray'}, {'x1': 360.1, 'y1': 50, 'x2': 379.1, 'y2': 50, 'color': 'yellow'}, {'x1': 379.2, 'y1': 50, 'x2': 412.2, 'y2': 50, 'color': 'orange'}, {'x1': 412.3, 'y1': 50, 'x2': 425.6, 'y2': 50, 'color': 'yellow'}, {'x1': 425.7, 'y1': 50, 'x2': 446.8, 'y2': 50, 'color': 'orange'}, {'x1': 441.3, 'y1': 50, 'x2': 446.8, 'y2': 50, 'color': 'gray'}]\n"
     ]
    }
   ],
   "source": [
    "print(spliced_d3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HW",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
